File: backend_go/internal/service/metrics.go
package service

import (
	"context"
	"log"
	"time"

	"mydashboard-backend/internal/models"
	"mydashboard-backend/internal/store"
)

type MetricsService struct {
	store     *store.Store
	simulator *Simulation
}

func NewMetricsService(store *store.Store, simulator *Simulation) *MetricsService {
	return &MetricsService{
		store:     store,
		simulator: simulator,
	}
}

func (s *MetricsService) Latest(ctx context.Context) (models.Metrics, error) {
	metrics, err := s.store.LatestMetrics(ctx)
	if err != nil {
		return models.Metrics{}, err
	}
	if metrics.CreatedAt.IsZero() {
		metrics = defaultMetrics()
		if err := s.store.InsertMetricsAt(ctx, metrics); err != nil {
			log.Printf("seed metrics failed: %v", err)
		}
	}
	return metrics, nil
}

func (s *MetricsService) Trend(ctx context.Context, window int) ([]models.Metrics, error) {
	points, err := s.store.Trend(ctx, window)
	if err != nil {
		return nil, err
	}
	if len(points) == 0 {
		points = seedTrendMetrics()
		for _, point := range points {
			if err := s.store.InsertMetricsAt(ctx, point); err != nil {
				log.Printf("seed trend failed: %v", err)
				break
			}
		}
	}
	return points, nil
}

func (s *MetricsService) Simulate(ctx context.Context) (models.Metrics, error) {
	metrics, err := s.store.LatestMetrics(ctx)
	if err != nil {
		return models.Metrics{}, err
	}
	if metrics.CreatedAt.IsZero() {
		metrics = defaultMetrics()
	}
	next := s.simulator.NextMetrics(metrics)
	if err := s.store.InsertMetrics(ctx, next); err != nil {
		return models.Metrics{}, err
	}
	return next, nil
}

func (s *MetricsService) StartSimulation(ctx context.Context, metricEvery, insightEvery time.Duration, insights *InsightsService) {
	metricsTicker := time.NewTicker(metricEvery)
	insightTicker := time.NewTicker(insightEvery)
	defer metricsTicker.Stop()
	defer insightTicker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-metricsTicker.C:
			if _, err := s.Simulate(ctx); err != nil {
				log.Printf("simulate metrics failed: %v", err)
			}
		case <-insightTicker.C:
			metrics, err := s.store.LatestMetrics(ctx)
			if err != nil {
				continue
			}
			if metrics.CreatedAt.IsZero() {
				metrics = defaultMetrics()
			}
			if _, err := insights.GenerateAuto(ctx, metrics); err != nil {
				log.Printf("simulate insight failed: %v", err)
			}
		}
	}
}

func defaultMetrics() models.Metrics {
	return models.Metrics{
		Revenue:   4.82,
		Growth:    18.6,
		Sentiment: 72,
		Backlog:   128,
		CreatedAt: time.Now(),
	}
}

func seedTrendMetrics() []models.Metrics {
	base := defaultMetrics()
	var points []models.Metrics
	for i := 0; i < 12; i++ {
		value := 55 + float64(i)*1.8 + (float64(i)/1.8)*2.0
		points = append(points, models.Metrics{
			Revenue:   value / 10,
			Growth:    base.Growth,
			Sentiment: base.Sentiment,
			Backlog:   base.Backlog,
			CreatedAt: time.Now().Add(time.Duration(i-12) * time.Minute),
		})
	}
	return points
}

File: backend/app/services/metrics.py
from __future__ import annotations

from datetime import datetime, timedelta
from typing import List

from ..models import Metrics
from ..store import Store
from .simulation import Simulation


class MetricsService:
    def __init__(self, store: Store, simulator: Simulation) -> None:
        self._store = store
        self._simulator = simulator

    def latest(self) -> Metrics:
        metrics = self._store.latest_metrics()
        if metrics is None:
            metrics = _default_metrics()
            try:
                self._store.insert_metrics_at(metrics)
            except Exception:
                pass
        return metrics

    def trend(self, window: int) -> List[Metrics]:
        points = self._store.trend(window)
        if not points:
            points = _seed_trend_metrics()
            for point in points:
                try:
                    self._store.insert_metrics_at(point)
                except Exception:
                    break
        return points

    def simulate(self) -> Metrics:
        metrics = self._store.latest_metrics() or _default_metrics()
        next_metrics = self._simulator.next_metrics(metrics)
        self._store.insert_metrics(next_metrics)
        return next_metrics


def _default_metrics() -> Metrics:
    return Metrics(
        revenue=4.82,
        growth=18.6,
        sentiment=72.0,
        backlog=128,
        created_at=datetime.now(),
    )


def _seed_trend_metrics() -> List[Metrics]:
    base = _default_metrics()
    points: List[Metrics] = []
    for i in range(12):
        value = 55 + float(i) * 1.8 + (float(i) / 1.8) * 2.0
        points.append(
            Metrics(
                revenue=value / 10,
                growth=base.growth,
                sentiment=base.sentiment,
                backlog=base.backlog,
                created_at=datetime.now() + timedelta(minutes=i - 12),
            )
        )
    return points

